docs/Building Environments in Aviary.ipynb

```diff
@@ -0,0 +1,434 @@
+    "\n",
+    "1. **Calculator(Expression)**: Return the result of a numerical expression.\n",
+    "2. **Check(Answer)**: Check if the answer is correct.\n",
+    "\n",
+    "The GSM8K environment can be represented as a Markov Decision Process $(\\mathcal{V}, \\mathcal{S}, \\mathcal{A}, \\mathcal{T}, R, \\gamma)$ as\n",
+    "\n",
+    "\n",
+    "- $\\mathcal{V}:$ The vocabulary - the English alphabet, together with punctuation symbols.\n",
+    "- $\\mathcal{S}:$ The state - GSM8K question, current step in the reasoning process.\n",
+    "- $\\mathcal{A}:$ The action - \\{Calculator, Check Answer\\}.\n",
+    "- $\\mathcal{T}:$ The deterministic transition function.\n",
+    "- $R:$ The reward function:\n",
+    "\n",
+    "    \\begin{cases} \n",
+    "    1 & \\text{if the action submits a correct answer}, \\\\\n",
+    "    -1 & \\text{if the action is an invalid tool call}, \\\\\n",
+    "    0 & \\text{otherwise}.\n",
+    "    \\end{cases}\n",
+    "\n",
+    "- $\\gamma:$ The discount factor\n",
+    "\n",
+    "Below we define the GSM8K environment in our language agent library called **Aviary**. The framework for defining an environment losely follows the template for reinforcement learning environments from Spinning up in Deep RL [2]. Custom implementations of Aviary environments should inherit from the `Environment` class and should implement the **reset**, and **step** methods in addition to defining functions for the tools the agent has access to."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8afcb60e-1b4b-48d4-9c78-a56794e75aa8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import contextlib\n",
+    "import json\n",
+    "from typing import Literal\n",
+    "\n",
+    "from pydantic import BaseModel, ConfigDict\n",
+    "\n",
+    "from aviary.core import (\n",
+    "    Environment,\n",
+    "    Message,\n",
+    "    TaskDataset,\n",
+    "    Tool,\n",
+    "    ToolRequestMessage,\n",
+    "    ToolResponseMessage,\n",
+    ")\n",
+    "\n",
+    "\n",
+    "class CalculatorEnvConfig(BaseModel):\n",
+    "    model_config = ConfigDict(extra=\"forbid\")\n",
+    "\n",
+    "    correct_reward: float = 1.0\n",
+    "    incorrect_reward: float = 0.0\n",
+    "    tool_failure_reward: float = -1.0\n",
+    "    tool_success_reward: float = 0.0\n",
+    "    rel_tol: float = 1e-4\n",
+    "\n",
+    "    done_on_failure: bool = True\n",
+    "\n",
+    "\n",
+    "class CalculatorEnv(Environment[None]):\n",
+    "    def __init__(\n",
+    "        self,\n",
+    "        problem_id: str,\n",
+    "        problem: str,\n",
+    "        answer: float,\n",
+    "        config: CalculatorEnvConfig | None = None,\n",
+    "    ):\n",
+    "        \"\"\"An environment for solving simple math problems using a calculator tool and a check answer tool.\n",
+    "\n",
+    "        Arguments:\n",
```

Comment:
[nit] Google Style uses `Args:` not `Arguments:`

Reply:
No idea how that ended up there, I think I copy/pasted from ChatGPT and didn't do due diligence haha